### PR TITLE
Adding options to translation averaging to support betweenfactor translation priors

### PR DESCRIPTION
**Pull request by akshay-krishnan**
These Point3 "between translations" (not using the term "relative translations" since this was already used for the direction measurements), can come for other sensors in a SLAM problem.

### CODE REVIEW

**Comment by dellaert** [comment]
Cool. But what's this?
image
Did you not start from develop?

**Comment by dellaert** [comment]
Two suggestions before I review:
- CI is failing, see what's up?
- I suggest you merge in develop

**Comment by akshay-krishnan** [comment]
I created a branch called ta-add-methods, but it looks like there was one already, and I started working on an existing branch. Pulled from develop now.

File: tests/testTranslationRecovery.cpp [hunk 1]
```diff
@@ -265,6 +265,74 @@ TEST(TranslationRecovery, ThreePosesWithZeroTranslation) {
...
+  poses.insert<Pose3>(1, Pose3(Rot3(), Point3(2, 0, 0)));
+  poses.insert<Pose3>(3, Pose3(Rot3(), Point3(1, -1, 0)));
+
+  auto relativeTranslations =
+      TranslationRecovery::SimulateMeasurements(poses, {{0, 1}, {0, 3}, {1, 3}});
+
+  TranslationRecoveryParams params;
+  std::vector<BinaryMeasurement<Point3>> betweenTranslations;
+  betweenTranslations.emplace_back(0, 1, Point3(2, 0, 0), noiseModel::Constrained::All(3, 1e2));
+  params.setBetweenTranslations(betweenTranslations);
```
Comment (dellaert) [inline tests/testTranslationRecovery.cpp:326]: Hmmm, I think these are not mere parameters, but *data*, and hence should be given as an (optional) argument to `run`

File: gtsam/sfm/TranslationRecovery.cpp [hunk 2]
```diff
@@ -71,36 +74,63 @@ TranslationRecovery::TranslationRecovery(
...
+                                                  edge.noiseModel());
+    }
+  }
+
   return graph;
 }
 
 void TranslationRecovery::addPrior(
-    const double scale, NonlinearFactorGraph *graph,
+    const double scale, const boost::shared_ptr<NonlinearFactorGraph> graph,
```
Comment (dellaert) [inline gtsam/sfm/TranslationRecovery.cpp:96]: I think `*` is right. In `.i` file use `@` to indicated an output parameter.

File: gtsam/sfm/TranslationRecovery.h [hunk 3]
```diff
@@ -28,6 +28,40 @@
 
 namespace gtsam {
 
+// Parameters for the Translation Recovery problem.
+class TranslationRecoveryParams {
```
Comment (dellaert) [inline gtsam/sfm/TranslationRecovery.h:32]: parameters are typically structs, with fields *without* `_` at end. This is now supported in wrapper, as well. That being said, I think betweenTranslations don't belong here.

File: gtsam/sfm/TranslationRecovery.h [hunk 4]
```diff
@@ -93,7 +125,8 @@ class TranslationRecovery {
    * @param graph factor graph to which prior is added.
    * @param priorNoiseModel the noise model to use with the prior.
    */
-  void addPrior(const double scale, NonlinearFactorGraph *graph,
+  void addPrior(const double scale,
```
Comment (dellaert) [inline gtsam/sfm/TranslationRecovery.h:128]: should remain *

File: gtsam/sfm/TranslationRecovery.h [hunk 5]
```diff
@@ -131,5 +165,25 @@ class TranslationRecovery {
...
+ private:
+  /**
+   * @brief Gets the key of the variable being optimized among multiple input
+   * variables that have the same translation.
+   *
+   * @param i key of input variable.
+   * @return Key of optimized variable - same as input if it does not have any
+   * zero-translation edges.
+   */
+  Key getUniqueKey(const Key i) const;
```
Comment (dellaert) [inline gtsam/sfm/TranslationRecovery.h:178]: I have no idea what this does. we can chat about it.

**Review by dellaert — COMMENTED** [review]
Cool, BUT, I think the params way is not right. We could have *only* betweenTranslations.

File: tests/testTranslationRecovery.cpp (hunk 1 above)
Comment (akshay-krishnan) [reply tests/testTranslationRecovery.cpp]: moved it to a `run` argument.

File: gtsam/sfm/TranslationRecovery.cpp (hunk 2 above)
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.cpp]: changed all shared_ptr back to *, this change was to wrap these methods, but later I decided not to wrap them.

File: gtsam/sfm/TranslationRecovery.h (hunk 3 above)
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.h]: Yeah, I would prefer struct too, I used the class for support in wrapper. Good to know that structs are supported now :)

File: gtsam/sfm/TranslationRecovery.cpp (hunk 2 above)
Comment (dellaert) [reply gtsam/sfm/TranslationRecovery.cpp]: This works in wrapper... Use @

File: gtsam/sfm/TranslationRecovery.h (hunk 5 above)
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.h]: If there are multiple input nodes that share the same translation, i.e have a zero unit translation edge between them, only one of them will be added to the factor graph. This function will return the value of the Key that was added to the graph for an input node. If there is no zero-translation edge, the output is the same as input. 

It is used each time we add the betweenFactors or any other priors, we need to find that "root" node that was added to the graph. I renamed the method, hopefully its more clear.

File: gtsam/sfm/sfm.i [hunk 6]
```diff
@@ -252,14 +262,21 @@ class MFAS {
...
-      const gtsam::BinaryMeasurementsUnit3&
-          relativeTranslations);  // default LevenbergMarquardtParams
-  gtsam::Values run(const double scale) const;
+      const gtsam::TranslationRecoveryParams& lmParams);
+  TranslationRecovery(const gtsam::BinaryMeasurementsUnit3&
+                          relativeTranslations);  // default params
+  gtsam::Values run(const gtsam::BinaryMeasurementsPoint3& betweenTranslations,
+                    const double scale) const;
+  // default empty betweenTranslations
+  // gtsam::Values run(const double scale) const;
```
Comment (akshay-krishnan) [inline gtsam/sfm/sfm.i:279]: @dellaert is there a way to wrap the method this way? using the default value of the first betweenTranslations argument?

Comment (dellaert) [reply gtsam/sfm/sfm.i:279]: Actually, I think run should have *both* `relativeTranslations` and `betweenTranslations`. The constructor should just take parameters, not data, so you can "run" multiple datasets.
If you agree, you could make scale the first argument and provide two declarations in the .i file:

```c++
gtsam::Values run(const double scale, const gtsam::BinaryMeasurementsUnit3& relativeTranslations) const;
gtsam::Values run(const double scale, const gtsam::BinaryMeasurementsUnit3& relativeTranslations, const gtsam::BinaryMeasurementsPoint3& betweenTranslations) const;
```

**Comment by akshay-krishnan** [comment]
That was not as simple as it should have been, due to the handling of nodes with zero-edges between them. But its done now, moved the relativeTranslations to `run()` . 
The handling of zero-edges now completely happens inside run(), using some helper functions that are not class methods.

**Review by dellaert — APPROVED** [review]

File: gtsam/sfm/TranslationRecovery.h [hunk 7]
```diff
@@ -28,6 +28,15 @@
 
 namespace gtsam {
 
+// Parameters for the Translation Recovery problem.
+struct TranslationRecoveryParams {
+  // LevenbergMarquardtParams for optimization.
+  LevenbergMarquardtParams lmParams;
+
+  // Initial values, random intialization will be used if not provided.
+  Values initial;
```
[the remark this answers is not in the document]
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.h]: updated that

File: gtsam/sfm/TranslationRecovery.cpp [hunk 8]
```diff
@@ -56,94 +56,148 @@ TranslationRecovery::TranslationRecovery(
...
+      removeSameTranslationNodes(relativeTranslations, sameTranslationDSFMap);
+  const std::vector<BinaryMeasurement<Point3>> nonzeroBetweenTranslations =
+      removeSameTranslationNodes(betweenTranslations, sameTranslationDSFMap);
+
+  // Create graph of translation factors.
+  NonlinearFactorGraph graph = buildGraph(nonzeroRelativeTranslations);
+
+  // Add global frame prior and scale (either from betweenTranslations or
+  // scale).
+  addPrior(nonzeroRelativeTranslations, scale, nonzeroBetweenTranslations,
```
[the remark this answers is not in the document]
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.cpp:186]: we do not add the scale prior, only a prior on the first translation to lie at (0, 0, 0). I updated the addPrior documention to be more clear about this.

**Pull request merged**